src/DataUtils.cpp

```diff
@@ -0,0 +1,211 @@
+		::MessageBoxA(nullptr, err_ss.str().c_str(), "Fatal Error", MB_ICONERROR | MB_OK);
+#else
+		// TODO: Add non-windows version.
+#endif
+		exit(0);
+	}
+
+	// Get list of required/optional attributes/children
+	// Sort
+	// Remove duplicates
+	// Rational for not using std:set:
+	// Profiled code takes average of 10 microseconds to complete.
+	auto requiredChildNames = NAS2D::split(requiredChildElements);
+	std::sort(std::begin(requiredChildNames), std::end(requiredChildNames));
+	requiredChildNames.erase(std::unique(std::begin(requiredChildNames), std::end(requiredChildNames)), std::end(requiredChildNames));
+
+	auto requiredAttributeNames = NAS2D::split(requiredAttributes);
+	std::sort(std::begin(requiredAttributeNames), std::end(requiredAttributeNames));
+	requiredAttributeNames.erase(std::unique(std::begin(requiredAttributeNames), std::end(requiredAttributeNames)), std::end(requiredAttributeNames));
+
+	auto optionalChildNames = NAS2D::split(optionalChildElements);
+	std::sort(std::begin(optionalChildNames), std::end(optionalChildNames));
+	optionalChildNames.erase(std::unique(std::begin(optionalChildNames), std::end(optionalChildNames)), std::end(optionalChildNames));
+
+	auto optionalAttributeNames = NAS2D::split(optionalAttributes);
+	std::sort(std::begin(optionalAttributeNames), std::end(optionalAttributeNames));
+	optionalAttributeNames.erase(std::unique(std::begin(optionalAttributeNames), std::end(optionalAttributeNames)), std::end(optionalAttributeNames));
+
+	// getChildElementNames
+	auto actualChildNames = [&element]() -> StringList {
+		NAS2D::StringList result{};
+		for (auto iter = element.firstChildElement();
+			 iter != nullptr;
+			 iter = iter->nextSiblingElement())
+		{
+			result.push_back(iter->value());
+		}
+		return result;
+	}(); // Immediately Invoked Initializing Lambda (IIIL)
+	std::sort(std::begin(actualChildNames), std::end(actualChildNames));
+	actualChildNames.erase(std::unique(std::begin(actualChildNames), std::end(actualChildNames)), std::end(actualChildNames));
+
+	// getAttributeNames
+	auto actualAttributeNames = [&element]() -> StringList {
+		NAS2D::StringList result{};
+		for (auto iter = element.firstAttribute();
+			 iter != nullptr;
+			 iter = iter->next())
+		{
+			result.push_back(iter->name());
+		}
+		return result;
+	}(); // Immediately Invoked Initializing Lambda (IIIL)
+	std::sort(std::begin(actualAttributeNames), std::end(actualAttributeNames));
+	actualAttributeNames.erase(std::unique(std::begin(actualAttributeNames), std::end(actualAttributeNames)), std::end(actualAttributeNames));
+
+	StringList extraChildElementNames{};
+	std::set_difference(
+		std::begin(actualChildNames), std::end(actualChildNames),
+		std::begin(requiredChildNames), std::end(requiredChildNames),
+		std::back_inserter(extraChildElementNames));
+	std::sort(std::begin(extraChildElementNames), std::end(extraChildElementNames));
```

Comment:
I may need to look at this section again later.

My thought here is the code is effectively doing a tree diff. There is an expected tree, and an actual tree. When doing a tree diff, if any required nodes are missing, it's an error. If any extra (optional) nodes are present, it (may be) a warning.

Reply:
There's no tree involved. There are two flat lists of values. A list of the unique names that actually appear in the file, and a list of unique names that are required *or* a list of names that are optional. The result `set_difference` returns a list of missing or extra names, depending on the order of the arguments (set_difference is not commutative). If an item is missing from the required set, that's an error. If an extra item is found that's not in the optional set, that's a warning.

Reply:
I meant conceptually, in the general case, of a deeply nested tree. I believe the function parameters here limits the number of levels it will actually test.

Reply:
Yes, it intentionally only tests the current level. This allows for different classes handling their own XML serialization by just passing in the current element.